src/platform/datapath_epoll.c

```diff
@@ -926,6 +926,29 @@ CxPlatSocketContextInitialize(
                 goto Exit;
             }
         }
+    } else if (SocketType == CXPLAT_SOCKET_TCP_LISTENER) {
+        //
+        // Set SO_REUSEADDR on listener sockets to avoid
+        // TIME_WAIT state on shutdown.
```

Comment:
I am a bit confused. The listener socket itself can go from listen state to timewait state? I thought a socket had to go through finwait1 and finwait2 to enter timewait state (which implicitly means only sockets who *initiate* graceful shutdown).

Reply:
Yes, on linux a listener socket itself can go into timewait. Its a bit weird. But it was extremely easy to reproduce.

Reply:
Mind sharing how it could be repro'd?

Just create a listener socket and have a connection accepted and kill both the accepted connection and the listener socket itself? Then bind will fail?

Reply:
Yup. Have the listener socket accept a connection, and then kill the listener side. The listener side then will not be able to restart due to a bind failure.

Reply:
I am really surprised linux has this weird behavior as it violates RFC793.

![image](https://github.com/user-attachments/assets/5185d339-bd71-4132-85ff-1e895c8eec59)

Reply:
Wouldn't the listener socket follow the left side of that tree down to TIME_WAIT?

Reply:
The socket that's accepted from the listener socket would follow the left side but the listener socket itself, in my understanding, would not. When you call accept() on a listener socket, you get a new socket and the listener itself should stay in listening state.

Reply:
It was a bit confusing to me too. libuv actually sets SO_REUSEADDR for both Client initiated sockets and listener sockets unconditionally, but at least for my use case I didn't need it set on the client side, and thats a little more invasive too.